changefeedccl: use single threaded dispatcher for event_processor

samiskin opened the pull request:
Epic: CRDB-19057

Previously event_processor would use a locked sink to emit to, resulting in many
worker threads contending on the same lock.

This change adds a channel with a buffer, as well as a single dispatcher loop to
pull events out of that channel to insert into the sink.  It also replaces the
mutex'd inflight counter with a waitgroup. These two changes allows us to avoid
an extra mutex lock on every emit, resulting in an increased throughput in
bottlenecked cases such as there being a large number of small events.

---
Running a cluster with:
- 3 gce nodes on `n2-standard-16`
- A TPCC fixture import of 800 warehouses
- A cloudstorage changefeed on all TPCC tables with no custom options specified

**Prior** to this PR, the backfill looked like this (green is the summed rate, the rest are per-node metrics):

<img width="811" alt="Screen Shot 2022-11-01 at 3 52 28 PM" src="https://user-images.githubusercontent.com/6236424/199325812-e9cbb153-91d0-4e48-bb18-3e0e4c2d287b.png">

We're initially bottlenecked on something else likely due to larger rows, however as the 3 nodes progress through their tables our byte throughput starts dropping as our emitted messages rate starts rising.  This indicates that we would be able to handle more data however something about the number of messages itself is limiting us.

**We peak at around 180 thousand messages per second**, taking around **20 minutes** to complete the backfill

---
With this PR, the backfill instead looks like this (**Ignore the initial spike**, that was when I accidentally had an old changefeed running after restarting the cluster):
<img width="717" alt="Screen Shot 2022-11-01 at 3 55 22 PM" src="https://user-images.githubusercontent.com/6236424/199326342-58f07c30-507e-4780-bb47-852e43c6e49b.png">

**We peak at around 230 thousand messages per second**, taking around **16.5 minutes** to complete the backfill.

Now as we get to the point later in the scan where our byte throughput dropped, it doesn't drop as much and the emitted messages is able to rise 27% higher than it did previously.  

---

Release note (performance improvement): minor improvement to maximum
changefeed throughput.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91059)
<!-- Reviewable:end -->

samiskin committed:
changefeedccl: use single threaded dispatcher for event_processor

Previously event_processor would use a locked sink to emit to, resulting
in many worker threads contending on the same lock.

This change adds a channel with a buffer, as well as a single dispatcher
loop to pull events out of that channel to insert into the sink.  It
also replaces the mutex'd inflight counter with a waitgroup. These two
changes allows us to avoid an extra mutex lock on every emit, resulting
in an increased throughput in bottlenecked cases such as there being a
large number of small events.

Release note (performance improvement): minor improvement to maximum
changefeed throughput.

samiskin commented:
Wanted to have a roachtest for having a repeatable performance test but will instead make that its own PR.

miretskiy commented on pkg/ccl/changefeedccl/event_processing.go:
```diff
@@ -368,14 +387,28 @@ func (c *kvEventToRowConsumer) ConsumeEvent(ctx context.Context, ev kvevent.Even
+		alloc:   a,
+	}
+	if c.sinkCh != nil {
+		c.sinkCh <- e
```
you should select with ctx.Done case.

miretskiy commented on pkg/ccl/changefeedccl/event_processing.go:
```diff
@@ -368,14 +387,28 @@ func (c *kvEventToRowConsumer) ConsumeEvent(ctx context.Context, ev kvevent.Even
+			return err
+		}
+
+		if log.V(3) {
```
might keep this log for both branch.

miretskiy commented on pkg/ccl/changefeedccl/event_processing.go:
```diff
@@ -622,9 +667,6 @@ func (c *parallelEventConsumer) Frontier() hlc.Timestamp {
 func (c *parallelEventConsumer) Close() error {
-	// Signal the done channel and wait for all workers to finish.
-	// If an error occurred, at least one worker will return an error, so
-	// it will be returned by c.g.Wait().
```
any particular reason to remove this comment?

miretskiy commented on pkg/ccl/changefeedccl/event_processing.go:
```diff
@@ -368,14 +387,28 @@ func (c *kvEventToRowConsumer) ConsumeEvent(ctx context.Context, ev kvevent.Even
+		updated: schemaTimestamp,
+		mvcc:    mvccTimestamp,
+		alloc:   a,
+	}
```
this sinkEvent will definitely result in an allocation (which escapes to the heap)...  It might make sense to have an sync.Pool of *sinkEvent so that those can be reused...

samiskin replied:
Done

miretskiy commented on pkg/ccl/changefeedccl/changefeedbase/settings.go:
```diff
@@ -252,3 +252,13 @@ var EventConsumerWorkerQueueSize = settings.RegisterIntSetting(
+		"which can be buffered by the dispatcher",
+	int64(util.ConstantWithMetamorphicTestRange("changefeed.event_consumer_dispatcher_buffer_size", 64, 0, 64)),
+	settings.NonNegativeInt,
+).WithPublic()
```
this strikes me as a very, very low level setting; almost certainly not "WithPublic" -- it might even just be a metamorphic constant...

samiskin replied:
Removed `WithPublic`, it just came along since I copied the similar nprocs setting.

miretskiy commented on pkg/ccl/changefeedccl/event_processing.go:
```diff
@@ -495,6 +539,32 @@ func (c *parallelEventConsumer) getBucketForEvent(ev kvevent.Event) int64 {
 	return int64(c.hasher.Sum32()) % c.numWorkers
 }
 
+func (c *parallelEventConsumer) dispatcher(ctx context.Context, sink EventSink) error {
```
nit: (well, maybe not a nit, but a request) I got confused initially thinking that dispatcher dispatches the work to the workers... where as in reality, it is the workers that emit to this dispatcher which then writes to the underlying sink -- thus ensuring sink usage is thread safe.  I think it would be clearer if this function was renamed emitRowWorker -- and, while at it, a comment added along the lines of what I typed above.

samiskin replied:
Renamed to `emitRowWorker` and I added a comment

miretskiy commented on pkg/ccl/changefeedccl/event_processing.go:
```diff
@@ -368,14 +387,28 @@ func (c *kvEventToRowConsumer) ConsumeEvent(ctx context.Context, ev kvevent.Even
+		updated: schemaTimestamp,
+		mvcc:    mvccTimestamp,
+		alloc:   a,
+	}
```
we don't have to allocate sinkEvent if sinkCh == nil, do we?

miretskiy commented on pkg/ccl/changefeedccl/changefeedbase/settings.go:
```diff
@@ -252,3 +252,13 @@ var EventConsumerWorkerQueueSize = settings.RegisterIntSetting(
+	"changefeed.event_consumer_dispatcher_buffer_size",
+	"if changefeed.event_consumer_workers is enabled, this setting sets the maxmimum number of events"+
+		"which can be buffered by the dispatcher",
+	int64(util.ConstantWithMetamorphicTestRange("changefeed.event_consumer_dispatcher_buffer_size", 64, 0, 64)),
```
I would like to see the same benchmark run with this setting set to 0.
The reason why I want to do this is that I want to test a hypothesis that with size of 0 performance should be almost identical to the old version.  Basically, this gives each worker a bit of buffer space so that they don't have to block on mutex  (to emit to sink) -- but with this set to 0, I suspect performance should be almost identical.

samiskin replied:
Interestingly enough with it set to 0 it is _much_ slower, in my `order_line` test doing 235k messages per second compared to the 300k of master.  This is with the only difference being that we're sending it to the worker and the worker is picking it off.    Im surprised that just sending it over a channel to another worker compared to lock/unlock makes such a huge difference.

jayshrivastava reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91059)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy and @samiskin)

___
*[`pkg/ccl/changefeedccl/event_processing.go` line 613 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91059#-NFsiENG5oOp50nJaU-5:-NFsiENG5oOp50nJaU-6:b-hrin1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8efdf3e707ed588d388277089ea980b916cc0a0/pkg/ccl/changefeedccl/event_processing.go#L613)):*
> ```Go
> 			if err := consumer.ConsumeEvent(ctx, e); err != nil {
> 				c.setWorkerError(err)
> 				c.decInFlight()
> ```

I don't think we need to call `decInFlight` here.
___
*[`pkg/ccl/changefeedccl/event_processing.go` line 623 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91059#-NFsjH9WEYEfSAhAO1mu:-NFsjH9WEYEfSAhAO1mv:b-m3zd7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8efdf3e707ed588d388277089ea980b916cc0a0/pkg/ccl/changefeedccl/event_processing.go#L623)):*
> ```Go
> 	c.inFlightWg.Add(1)
> 	atomic.AddInt32(&c.inFlightCounter, 1)
> 	c.metrics.ParallelConsumerInFlightEvents.Update(int64(c.inFlightCounter))
> ```

I think `c.metrics.ParallelConsumerInFlightEvents.Update(int64(atomic.AddInt32(&c.inFlightCounter, 1)))` would be better here so you atomically update the counter and read its new value.



_Code quote:_
```Go
c.metrics.ParallelConsumerInFlightEvents.Update(int64(c.inFlightCounter))
```
___
*[`pkg/ccl/changefeedccl/event_processing.go` line 642 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/91059#-NFsk5YL8wSFeFfHKjJT:-NFsk5YL8wSFeFfHKjJU:b-45onv9) ([raw file](https://github.com/cockroachdb/cockroach/blob/a8efdf3e707ed588d388277089ea980b916cc0a0/pkg/ccl/changefeedccl/event_processing.go#L642)):*
> ```Go
> // Flush flushes the consumer by blocking until all events are consumed,
> // or until there is an error.
> func (c *parallelEventConsumer) Flush(ctx context.Context) error {
> ```

This is nice!


<!-- Sent from Reviewable.io -->

samiskin commented:
Another data point with tpcc of 800 only running on the `order_line` table.  Time to completion was 18% faster, maximum throughput was 11% higher
<img width="550" alt="Screenshot 2022-11-28 at 10 51 34 AM" src="https://user-images.githubusercontent.com/6236424/204322192-64736560-ab36-4915-a554-3833353536f4.png">

miretskiy commented:
@samiskin -- are the previous comments all addressed?

samiskin commented:
Yep I just need to do the "set the channel buffer size to 1 and compare to master" test, I'll add individual replies soon

samiskin commented on pkg/ccl/changefeedccl/event_processing.go:
```diff
@@ -626,5 +725,11 @@ func (c *parallelEventConsumer) Close() error {
+	waitErr := c.g.Wait()
+
+	// ensure goroutines blocked on inFlightWg.Wait() can exit
+	for c.inFlightCounter > 0 {
+		c.decInFlight()
+	}
```
This felt a little hacky but I wanted to avoid requiring that we be extra careful that _every_ possible path of completing work on an event has a matching `dec`, it felt too dangerous to be relied on even if all races were fixed here.

samiskin reviewed (commented):
@miretskiy

shermanCRL commented:
Ready to merge?

samiskin commented:
The plan now is just to leverage the event processor parallelism with a dumb sink that works in parallel, such that we don't need to merge parallel processing into a single thread only to re-parallelize it again afterwards, so its more that this PR is being abandoned.

shermanCRL commented:
Close this? Achieved by the other work?

miretskiy commented:
@samiskin close this PR?

samiskin commented:
Closing, we're basically getting this via sinks that don't need the safeSink wrapper now (webhook and pubsub)
